flexget/utils/archive.py

```diff
@@ -0,0 +1,186 @@
+"""
+Utilities for handling RAR and ZIP archives
+
+Provides wrapper archive and exception classes to simplify
+archive extraction
+"""
+from __future__ import unicode_literals, division, absolute_import
+from builtins import *  # pylint: disable=unused-import, redefined-builtin
+
+import zipfile
+import os
+import re
+import shutil
+import logging
+
+try:
+    import rarfile
+except ImportError:
+    rarfile = None
+
+
+log = logging.getLogger('archive')
+
+class ArchiveError(Exception):
+    """Base exception for archive"""
+    pass
+class NeedRarFile(ArchiveError):
+    """Exception to be raised when rarfile module is missing"""
+    pass
+class BadArchive(ArchiveError):
+    """Wrapper exception for BadZipFile and BadRarFile"""
+    pass
+class NeedFirstVolume(ArchiveError):
+    """Wrapper exception for rarfile.NeedFirstVolume"""
+    pass
+
+def rarfile_set_tool_path(config):
+    """
+    Manually set the path of unrar executable if it can't be resolved from the
+    PATH environment variable
+    """
+    unrar_tool = config['unrar_tool']
+
+    if unrar_tool:
+        if not rarfile:
+            log.error('rar_tool specified with no rarfile module installed.')
+        else:
+            rarfile.UNRAR_TOOL = unrar_tool
+            log.debug('Set RarFile.unrar_tool to: %s', unrar_tool)
+
+def rarfile_set_path_sep(separator):
+    """
+    Set the path separator on rarfile module
+    """
+    if rarfile:
+        rarfile.PATH_SEP = separator
+
+class Archive(object):
+    """
+    Base archive class. Assumes an interface similar to
+    zipfile.ZipFile or rarfile.RarFile
+    """
+    def __init__(self, archive_object, path):
+        self.path = path
+        self.archive = archive_object(self.path)
+        log.debug('Successfully opened RAR: %s', self.path)
+
+    def close(self):
+        """Release open resources."""
+        self.archive.close()
+
+    def delete(self):
+        """Delete the volumes that make up this archive"""
+        volumes = self.volumes()
+        self.close()
+
+        for volume in volumes:
+            log.debug('Deleting volume: %s', volume)
+            os.remove(volume)
```

Comment:
What should I do instead, add another exception to raise in its stead? Obviously I still need to let the calling code know that something went wrong.